### PR TITLE
[SETL-113] Check deliveries before computing

### PR DESCRIPTION
**Pull request by JorisTruong**
#113

### CODE REVIEW

**Comment by JorisTruong** [comment]
Comparing string type and string deliveryId of:
1. Input Deliverables in the Pipeline and factories output that can be used as an input Deliverable
2. Needed Deliverable in Factories

Bug with ```autoLoad = true```:
```SparkRepository[T] ``` doesn't match with ```Dataset[T]```

File: src/main/scala/com/jcdecaux/setl/workflow/Pipeline.scala [hunk 1]
```diff
@@ -284,6 +285,34 @@ class Pipeline extends Logging
...
+    val pipelineDeliverables = _deliverableDispatcher
+      .getRegistry
+      .values
+      .map(v => s"${ReflectUtils.getPrettyName(v.runtimeType)} ${v.deliveryId}")
+      .toSet
+
+    // Find middle factories output that can be used as a deliverable
+    val stagesOutput = stages
+      .flatMap(s => s.factories)
+      .dropRight(1)
```
[the remark this answers is not in the document]
Comment (JorisTruong) [reply src/main/scala/com/jcdecaux/setl/workflow/Pipeline.scala]: Is this case possible ?
* A pipeline has 2 stages.
* The first stage has one Factory.
* The second stage has two Factories.
* In this last stage, the second Factory needs, as input, the output of the first Factory.

In this case, we should only drop the "last" factory, shouldn't we ? Is the output of the first Factory passed in the next Stage or the next Factory ?

[the remark this answers is not in the document]
Comment (qxzzxq) [reply src/main/scala/com/jcdecaux/setl/workflow/Pipeline.scala]: All the factories within the same stage will be executed in parallel. so they can't depend on each other. If factory B depends on factory A, then the factory A must be in an earlier stage than the one of B

File: src/main/scala/com/jcdecaux/setl/transformation/FactoryDeliveryMetadata.scala [hunk 2]
```diff
@@ -38,7 +38,7 @@ private[setl] case class FactoryDeliveryMetadata(factoryUUID: UUID,
    *
    * @return
    */
-  def getFactoryInputs: List[FactoryInput] = argTypes.map(tp => FactoryInput(tp, producer, id))
+  def getFactoryInputs(consumer: Class[_ <: Factory[_]]): List[FactoryInput] = argTypes.map(tp => FactoryInput(tp, producer, id, autoLoad, optional, consumer))
```
Comment (qxzzxq) [inline src/main/scala/com/jcdecaux/setl/transformation/FactoryDeliveryMetadata.scala]: Does the argument `consumer` always equal to the class of this factory? In this case maybe we add a field called `factoryClass` which is the class object of the factory that the delivery belongs to, for example: 
```scala
/**
 * DeliverySetterMetadata contains information of the @Delivery annotated method, including the name,
 * argument types, the producer and optional
 *
 * @param factoryUUID UUID of factory
 * @param factoryClass class of factory that has this delivery
 * @param symbol      symbol of the method
 * @param deliverySetter      either a field or a method
 * @param argTypes    type of each argument
 * @param producer    the producer class for this delivery
 * @param optional    true if this delivery is optional
 * @param autoLoad    true if this delivery should be automatically loaded
 * @param condition    string, filter condition that wil be applied during the auto-load
 * @param id    string, id of this delivery
 */
private[setl] case class FactoryDeliveryMetadata(factoryUUID: UUID,
                                                 factoryClass: Class[_ <: Factory[_]],
                                                 symbol: runtime.universe.Symbol,
                                                 deliverySetter: Either[Field, Method],
                                                 argTypes: List[runtime.universe.Type],
                                                 producer: Class[_ <: Factory[_]],
                                                 optional: Boolean,
                                                 autoLoad: Boolean = false,
                                                 condition: String = "",
                                                 id: String = Deliverable.DEFAULT_ID) {
...
}
```

And then we can keep the argument of this method empty. It's maybe more understandable

File: src/main/scala/com/jcdecaux/setl/transformation/FactoryInput.scala [hunk 3]
```diff
@@ -17,4 +17,7 @@ import scala.reflect.runtime
  */
 private[setl] case class FactoryInput(override val runtimeType: runtime.universe.Type,
                                       producer: Class[_],
-                                      deliveryId: String = Deliverable.DEFAULT_ID) extends HasType
+                                      deliveryId: String = Deliverable.DEFAULT_ID,
+                                      autoLoad: Boolean,
+                                      optional: Boolean,
+                                      consumer: Class[_ <: Factory[_]]) extends HasType
```
Comment (qxzzxq) [inline src/main/scala/com/jcdecaux/setl/transformation/FactoryInput.scala]: Following the comment in `FactoryDeliveryMetadata`, maybe it's more understandable if we rename the filed `consumer` to `factoryClass`

File: src/main/scala/com/jcdecaux/setl/util/CheckDeliverable.scala [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package com.jcdecaux.setl.util
+
+import com.jcdecaux.setl.transformation.Factory
+
+case class CheckDeliverable(
```
Comment (qxzzxq) [inline src/main/scala/com/jcdecaux/setl/util/CheckDeliverable.scala]: Maybe `ExpectedDeliverable`?

**Pull request merged**